cloud/pkg/apis/edgeapplication/v1alpha1/type.go

```diff
@@ -0,0 +1,221 @@
+
+	// Component is part of image name.
+	// Basically we presume an image can be made of '[registry/]repository[:tag]'.
+	// The registry could be:
+	// - k8s.gcr.io
+	// - fictional.registry.example:10443
+	// The repository could be:
+	// - kube-apiserver
+	// - fictional/nginx
+	// The tag cloud be:
+	// - latest
+	// - v1.19.1
+	// - @sha256:dbcc1c35ac38df41fd2f5e4130b32ffdb93ebae8b3dbe638c23575912276fc9c
+	//
+	// +kubebuilder:validation:Enum=Registry;Repository;Tag
+	// +required
+	Component ImageComponent `json:"component"`
+
+	// Operator represents the operator which will apply on the image.
+	// +kubebuilder:validation:Enum=add;remove;replace
+	// +required
+	Operator OverriderOperator `json:"operator"`
+
+	// Value to be applied to image.
+	// Must not be empty when operator is 'add' or 'replace'.
+	// Defaults to empty and ignored when operator is 'remove'.
+	// +optional
+	Value string `json:"value,omitempty"`
+}
+
+// ImagePredicate describes images filter.
+type ImagePredicate struct {
+	// Path indicates the path of target field
+	// +required
+	Path string `json:"path"`
+}
+
+// ImageComponent indicates the components for image.
+type ImageComponent string
+
+// OverriderOperator is the set of operators that can be used in an overrider.
+type OverriderOperator string
+
+// These are valid overrider operators.
+const (
+	OverriderOpAdd     OverriderOperator = "add"
+	OverriderOpRemove  OverriderOperator = "remove"
+	OverriderOpReplace OverriderOperator = "replace"
+)
+
+// Manifest represents a resource to be deployed on managed node groups.
+type Manifest struct {
+	// +kubebuilder:pruning:PreserveUnknownFields
+	runtime.RawExtension `json:",inline"`
+}
+
+// EdgeAppStatus defines the observed state of EdgeApplication.
+type EdgeAppStatus struct {
+	// Conditions contain the different condition statuses for this work.
+	// Valid condition types are:
+	// 1. Applied represents workload in EdgeApplication is applied successfully on a managed node groups.
+	// 2. Progressing represents workload in EdgeApplication is being applied on a managed node groups.
+	// 3. Available represents workload in EdgeApplication exists on the managed node groups.
+	// 4. Degraded represents the current state of workload does not match the desired
+	// state for a certain period.
+	// +optional
+	Conditions []metav1.Condition `json:"conditions,omitempty"`
+
+	// ManifestStatuses contains running status of manifests in spec.
+	// +optional
+	ManifestStatuses []ManifestStatus `json:"manifestStatuses,omitempty"`
+}
+
+// ManifestStatus contains running status of a specific manifest in spec.
+type ManifestStatus struct {
+	// Identifier represents the identity of a resource linking to manifests in spec.
+	// +required
+	Identifier ResourceIdentifier `json:"identifier"`
+
+	// Status reflects running status of current manifest.
+	// +kubebuilder:pruning:PreserveUnknownFields
+	// +optional
+	Status *runtime.RawExtension `json:"status,omitempty"`
```

Comment:
Will we copy the status filed of manifests as a whole into the ManifestStatus here?

Reply:
yes